### DISCUSSION TITLE
Added: C# Language Design Notes for May 31, 2017

### DISCUSSION THREAD

**Discussion by MadsTorgersen** [dotnet/csharplang#691] [General]
[C# Language Design Notes for May 31, 2017](https://github.com/dotnet/csharplang/blob/master/meetings/2017/LDM-2017-05-31.md)

1. Default interface members: overriding or implementing?
2. Downlevel poisoning of ref readonly in signatures
3. Extension methods with ref this and generics
4. Default in operators

Please discuss.

**Comment by Thaina** [comment]
> Extension methods with ref this and generics

I think we can safe to allow constraint interface only with `struct`  constraint

```C#
void A<T>(ref this T data) where T : struct,InterfaceA,InterfaceB,InterfaceC
{
    DoSomethingT;
}
```

**Comment by JeroenBos** [comment]
Regarding the weird syntax like
`base(I3).(I1.M1)<string>(1, 2, 3); // Call I3's implementation of I1's M1`

Isn't `base(I3).M<...>(...)` already unambiguous? 

And why has been chosen for the syntax `base(IType).M`? That's not idiomatic C#. I know this requires more parentheses, but I prefer `((IType)base).M`. It doesn't require new syntax.

**Comment by sirgru** [comment]
@JeroenBos 
> Isn't  base(I3).M<...>(...) already unambiguous?

I was just thinking the same thing. Also, maybe it's in the other proposals, but what is the result of `base(T4)` where `interface I4: I2, I3 { ... }`? This seems a little ambiguous. 
As far calling `base(T)` this has been discussed here https://github.com/dotnet/csharplang/blob/master/meetings/2017/LDM-2017-06-14.md and I personally prefer this style over the alternatives.

**Comment by svick** [comment]
@JeroenBos 

> And why has been chosen for the syntax `base(IType).M`? That's not idiomatic C#.

Of course it's not idiomatic in the current version of C#. It's doing something that's not possible in the current version of the language, so it pretty much can't use existing syntax.

> I know this requires more parentheses, but I prefer `((IType)base).M`. It doesn't require new syntax.

It does. `base` is not an expression on its own, according to the specification, so the current syntax does not allow you to write that code and allowing it would be introducing new syntax. (Though `base` is an expression according to Roslyn.)

**Comment by alrz** [comment]
Using a cast (although, just syntactically) doesn't make sense, because virtual dispatch doesn't care about static type of an expression, same goes with casting `this`. So you're not communicating what is happening here with a cast.
